api/krusty/gvknpatch_test.go

```diff
@@ -0,0 +1,633 @@
+spec:
+  template:
+    spec:
+      containers:
+        - name: nginx
+          image: nginx
+`)
+
+	th.WriteF("apps/patch.yaml", `
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: new-name
+`)
+
+	th.WriteK("apps", `
+resources:
+- deployment.yaml
+
+patches:
+- path: patch.yaml
+  target:
+    kind: Deployment
+`)
+
+	options := th.MakeDefaultOptions()
+	options.ChangeId = true
+
+	// kind should become `StatefulSet`
+	// name should become `new-name`
+	m := th.Run("apps", options)
+	th.AssertActualEqualsExpected(m, `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: old-name
+spec:
+  template:
+    spec:
+      containers:
+      - image: nginx
+        name: nginx
+`)
+}
+
+// https://github.com/kubernetes-sigs/kustomize/issues/3280
+// Should be able to refer to a resource with either its
+// original GVKN or its current one
+func TestPatchOriginalName(t *testing.T) {
+	th := kusttest_test.MakeHarness(t)
+
+	th.WriteF("base/deployment.yaml", `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: old-name
+spec:
+  template:
+    spec:
+      containers:
+        - name: nginx
+          image: nginx
+`)
+	th.WriteF("base/patch.yaml", `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: new-name
+`)
+	th.WriteK("base", `
+resources:
+- deployment.yaml
+
+patches:
+- path: patch.yaml
+  target:
+    kind: Deployment
+`)
+
+	th.WriteK("overlay", `
+resources:
+- ../base
+patchesStrategicMerge:
+- depPatch.yaml
+`)
+	th.WriteF("overlay/depPatch.yaml", `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: old-name
+spec:
+  replicas: 999
+`)
+
+	options := th.MakeDefaultOptions()
+	options.ChangeId = true
+
+	// name should become `new-name`
+	m := th.Run("overlay", options)
+	th.AssertActualEqualsExpected(m, `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: old-name
+spec:
+  replicas: 999
+  template:
+    spec:
+      containers:
+      - image: nginx
+        name: nginx
+`)
+}
+
+func TestPatchNewName(t *testing.T) {
+	th := kusttest_test.MakeHarness(t)
+
+	th.WriteF("base/deployment.yaml", `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: old-name
+spec:
+  template:
+    spec:
+      containers:
+        - name: nginx
+          image: nginx
+`)
+	th.WriteF("base/patch.yaml", `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: new-name
+`)
+	th.WriteK("base", `
+resources:
+- deployment.yaml
+
+patches:
+- path: patch.yaml
+  target:
+    kind: Deployment
+`)
+
+	th.WriteK("overlay", `
+resources:
+- ../base
+patchesStrategicMerge:
+- depPatch.yaml
+`)
+	th.WriteF("overlay/depPatch.yaml", `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: new-name
+spec:
+  replicas: 999
+`)
+
+	options := th.MakeDefaultOptions()
+	options.ChangeId = true
+
+	// depPatch cannot find target with the name `new-name`
+	// because base/patch.yaml can't yet edit the name
+	err := th.RunWithErr("overlay", options)
+	if err == nil {
+		t.Fatalf("expected an error")
+	}
```

Comment:
```
assert.Error(t, th.RunWithErr("overlay", options))
```
Also, add ` "github.com/stretchr/testify/assert" ` to imports.

I resisted using this package too long.  It's pretty solid.